Simplify and fix `Projection`'s getter functions

Flarkk opened the pull request:
Fixes https://github.com/godotengine/godot/issues/90427
Fixes https://github.com/godotengine/godot/issues/93683

This PR simplifies and fixes a few functions of struct `Projection` :
- Fix `get_aspect()` and `get_lod_multiplier()` with camera in `PROJECTION_FRUSTUM mode. The latter now has a uniform behavior no matter the camera mode (perspective, frustum and orthographic). Besides it was broken in frustum mode, it also used to decimate meshes twice slower in orthographic than in perspective mode.
- Fix and document `get_pixels_per_meter()`. **Note** : this function has a clumsy definition and is not used anywhere. It's a good candidate for deprecation.
- Optimize `get_z_far()`, `get_z_near()`, `get_viewport_half_extents()`, `get_far_plane_half_extents()` and `get_fov()`
- Add (more) unit test to the above functions

Test project for LOD multiplier issues : [lod.zip](https://github.com/user-attachments/files/18065186/lod.zip)

Flarkk committed:
Simplify and fix `Projection`'s getter functions

Ivorforce commented on core/math/projection.cpp:
```diff
@@ -919,53 +858,31 @@ Projection::operator String() const {
-	if ((matrix[8] == 0) && (matrix[9] == 0)) {
-		return Math::rad_to_deg(Math::acos(Math::abs(right_plane.normal.x))) * 2.0;
+	if (columns[2][0] == 0) {
+		return Math::rad_to_deg(2 * Math::atan(1.0 / columns[0][0]));
```
Use `atan2` for `atan(a / b)` :)

Flarkk replied:
Oh yeah good catch

Flarkk commented:
Replaced `atan(a / b)` by `atan2(a, b)`. Thanks @Ivorforce

Ivorforce reviewed (commented):
This PR is very interesting to me, considering it could eliminate a large number of computations in the rendering pipeline.
However, I'm also a bit concerned about the reduction in complexity. I _think_ you're making a few assumptions that may not be true in nonstandard setups. 

May I ask, how did you arrive at the simplifications in this PR?

Flarkk commented on core/math/projection.cpp:
```diff
@@ -402,82 +402,21 @@ void Projection::set_frustum(real_t p_size, real_t p_aspect, Vector2 p_offset, r
-	new_plane.normalize();
-
-	return new_plane.d;
+	return -(columns[3][3] - columns[3][2]) / (columns[2][2] - columns[2][3]);
```
The far plane's homogeneous coordinate (`Vector4`) is row[3] - row[2]  (see previous implementation).
Zfar is the intercept of the plane, namely the `w` component of the above vector *after normalization*.

I'm making one assumption here that is the far plane is perpendicular to Z (i.e. both `x`  and `y` equal 0 in the above vector) which is the case in all current projection modes[^1].
[^1]: `x`  and `y`  may not be 0 anymore once we introduce oblique clipping planes. A specific code path might be needed then to perform a full normalization. That being said, Zfar kind of looses all its meaning with oblique planes, so I'm not even sure it would be relevant at all.

Under those conditions, normalizing the vector is just doing `w / abs(z)`, which is exacty what's done in the new implementation.

I made sure all unit tests still pass (and added new ones to cover all projection modes).
Of course it's not a guarantee that the code is bug free, but I'm pretty confident on this one.
If you can add more tests to cover the edge cases you may have in mind, please do.

Flarkk commented:
> I _think_ you're making a few assumptions that may not be true in nonstandard setups.

> May I ask, how did you arrive at the simplifications in this PR?

Sure thing. Here we are :

### Assumptions

There are 3 general assumptions, and only really 1 new :
1. the matrix is a projection matrix and not an arbitrary transformation - *already taken in previous implementation*
2. near and far planes are rectangles - *already taken in previous implementation*
3. near and far planes are orthogonal to z-axis - **the new one**, allows simplying `get_z_near()` and `get_z_far()`

All 3 are always true in our current projection modes (ORTHOGONAL, PERSPECTIVE, FRUSTUM).
Note that no.3 may not be true anymore when we introduce oblique near / far planes. However with oblique planes the very notions of znear and zfar become nonsensical, so it may not be an issue.

### Rationale
- `get_z_near()` and `get_z_far()` : see [my comment](https://github.com/godotengine/godot/pull/100209#discussion_r1917157513)
- `get_aspect()` : the matrix diagonal's `Dx` and `Dy` values are by definition `k / extent.x` (resp. `k / extent.y`) with `extent.xy` the near plane rectangle's size. X:Y aspect is just the ratio `Dy / Dx`
- `get_viewport_half_extents()` and `get_far_plane_half_extents()` : we're looking for the vector `Vector4(x, y, -[zfar or znear], 1)` whose transformed clip-space coordinate is `Vector3(w, w, ±w)`. Assuming a symetrical frustum (if not these functions are irrelevant) this yields `x = w / Dx` and `y = w / Dx` with `w = -[zfar or znear] * m[2][3] + 1 * m[3][3]` as the direct consequence of the matrix multiplication.
- `get_fov()` : angle between center and right bound of the view frustum can be seen as the `atan` of viewport's right distance over near distance. I kept the two code paths as it saves one `atan2` for symmetrical frustums (most common case by far)
- `get_lod_multiplier()` : it's straightforward to notice that the current implementation returns `1.0f / (zn / width)` in PERSPECTIVE mode, which can be simplified into `2 / Dx`. In ORTHOGONAL mode however, it currently returns `get_viewport_half_extents()` which is `1 / Dx`. Both paths now behave the same which means meshes now decimate twice faster in ORTHOGONAL mode.
- `get_pixels_per_meter()` : I couldn't get what it is supposed to return (code and documentation are inconsistent, and the function is not used anywhere). I chose to return the most sensical thing to me, which is *the user input in pixels divided by the viewport's width measured in meters on the near plane*. When used with the viewport's pixel width as an input, it returns the viewport's resolution in pixels per meter

Ivorforce commented:
> Sure thing. Here we are :
> [lots of great explanations]

Awesome, thank you for the detailed breakdown!
Unfortunately, this PR is too far outside my own expertise to be confident to give you an approval, but your explanations make sense to me (and code wise, it's good too!).

I do have one request for the PR: Please document assumptions you make (or specifically, those you do not make) in the respective functions. It's a bit of an unfair ask, since they haven't been documented before, but I think it's essential to keep the code integrity intact for maintainers and future contributors.

Here's a rough example of what I mean:
```c++
real_t Projection::get_z_far() const {
	// NOTE: This assumes that:
	//             - The far plane is a rectangle orthogonal to z (incorrect for oblique clipping planes, which are not supported).
	//             - The matrix is a valid projection matrix.
 	return -(columns[3][3] - columns[3][2]) / (columns[2][2] - columns[2][3]);
}
```

This might be a bit verbose, based on the overarching assumptions of the `Projection` class itself, so feel free to tweak the granularity of the statements. But I do think they have to be there in some form, to protect it against regressions :)

Flarkk commented:
@Ivorforce just took a stab at it - see my last push.
I took this opportunity to also improve 2 minor things :
- removed the need for a negation in `get_z_far()` by incorporating the minus sign in the denominator
- changed `is_orthogonal()` to `return columns[2][3] == 0.0` instead of testing whether `columns[3][3] == 1.0`. This is slightly more generic as the latter could return `true` with some interpolated perspective/orthogonal matrices setups (which have both `columns[2][3] != 0 && columns[3][3] != 0`). These are not part of the currently allowed camera setups though, so this makes no difference right now.

Ivorforce commented:
Awesome!! This es exactly what I was looking for.
Somebody should probably double check the math, and functionality in bigger projects - but between your comments, explanations and unit tests i feel comfortable giving this a thumbs up. If we merge it early in the 4.5 cycle there should be enough time to verify everything works as expected.
Great job!

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/c2732ae4b6fa4e7b543c065bedcd92759f6db387), it works as expected.

I've tested this on various 3D demos and couldn't spot any issues.

Unit tests look good to me. I don't know if the math changes here are sound though, as I am not well-versed in this domain.

Repiteo commented on core/math/projection.cpp:
```diff
@@ -919,53 +872,45 @@ Projection::operator String() const {
+	// - the matrix is a projection across z-axis (i.e. is invertible and columns[0][1], [0][3], [1][0] and [1][3] == 0)
+	// - the projection plane is rectangular (i.e. columns[0][2] and [1][2] == 0 if columns[2][3] != 0)
+	real_t width = 2.0f * (-get_z_near() * columns[2][3] + columns[3][3]) / columns[0][0];
+	return int(p_for_pixel_width / width); // The cast to int should be removed at some point (kept for compatibility for now).
```
Does this cast actually do anything? The return type is already `int`, so I would assume this is redundant.

Flarkk replied:
Will remove it indeed.
What I meant in the comment is there is no reason this function returns int and not real_t, but I left it as is for now.

Repiteo commented:
Thanks!
